code/data_setup/image_cropping.py

```diff
@@ -0,0 +1,213 @@
+    cropped_area_x1 = min(box_for_crop.x_center + box_for_crop.width, img_coords.width)
+    cropped_area_y1 = min(box_for_crop.y_center + box_for_crop.height, img_coords.height)
+    cropped_area_coords = Coords(cropped_area_x0, cropped_area_y0,
+                                 cropped_area_x1, cropped_area_y1)
+
+    new_box_coords = get_new_box_coords(box_coords, cropped_area_coords,
+                                        output_img_width, output_img_height)
+    return cropped_area_coords, new_box_coords
+
+def read_bbox_data(raw_image_dir, parsed_bb_path='work/imagenet/parsed_bb.csv'):
+
+    # paper says they use .66, but they also have search area that is twice bounding
+    # box. So, 0.5 may be more appropriate.
+    max_box_frac_of_width =  0.66
+    max_box_frac_of_height = 0.66
+    images_successfully_downloaded = set(os.listdir('data/imagenet/images'))
+    bbox_df = (pd.read_csv('work/imagenet/parsed_bb.csv')
+                    .assign(box_height = lambda df: df.ymax - df.ymin,
+                            box_width = lambda df: df.xmax - df.xmin)
+                    .assign(box_frac_of_height = lambda df: df.box_height / df.height,
+                            box_frac_of_width = lambda df:  df.box_width / df.width)
+                    .query('filename in @images_successfully_downloaded')
+                    .query('box_frac_of_height < @max_box_frac_of_height')
+                    .query('box_frac_of_width < @max_box_frac_of_width'))
+    return bbox_df
+
+if __name__ == "__main__":
+    output_width = 256
+    output_height = 256
+    raw_image_dir = 'data/imagenet/images/'
+    img_metadata = read_bbox_data(raw_image_dir)
+
+    i = 0
+    for _, row in img_metadata.iterrows():
+        i += 1
+        if i < 10:
+            img = cv2.imread(raw_image_dir + row.filename)
+            original_img = img.copy()   # only used for visualization. will be removed
+            img_coords = Coords(0, 0, row.width, row.height)
+            box_coords = Coords(row.xmin, row.ymin, row.xmax, row.ymax)
+            cv2.rectangle(img,
+                          pt1=(box_coords.x0, box_coords.y0),
+                          pt2=(box_coords.x1, box_coords.y1),
+                          color=(255, 100, 0),
+                          thickness=2)
+            show_img("Original", img)
+            crop_coords, new_box_coords = get_random_crop_coords(box_coords,
+                                                                 img_coords,
+                                                                 output_img_width=output_width,
+                                                                 output_img_height=output_height)
+            print("------------")
+            print(img_coords)
+            print(box_coords)
+            print(crop_coords)
+            print(new_box_coords)
+            cropped_img = original_img[crop_coords.x0:crop_coords.x1,
```

Comment:
This was the error, and it's an error I've made many times before.

⚠️  You index into an image as `img[y0, x0, y1, x1]` instead of `img[x0, y0, x1, y1]`. Because the array is stored as rows, and a row number corresponds to a height.